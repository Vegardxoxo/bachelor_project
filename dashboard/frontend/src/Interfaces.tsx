// Interfaces used for the application
export interface LicensePoolData {
    application_name: string;
    freed_by_organization: string;
    details: Array<{
        id: number;
        freed_by_organization: string;
        application_name: string;
        date_added: string;
        family: string | null;
        family_version: string | null;
        family_edition: string | null;
        price: number;
        spc_id: number;
    }>;
}

export interface OwnOrgData {
    application_name: string;
    primary_user_full_name: string;
    primary_user_email: string;
    organization: string;
    computer_name: string;
    details: Array<{
        id: number;
        last_used: string;
        status: string;
        price: number;
    }>
}


export interface SoftwareData {
    id: number;
    application_name: string;
    computer_name: string;
    primary_user_full_name: string;
    primary_user_email: string;
    last_used: string;
    total_minutes: number;
    active_minutes: number;

}

export interface Count {
    total_licenses: number,
    active_licenses: number,
    never_used: number,
    unused_licenses: number,
    available_licenses: number,

}

export interface UserInformation {
    primary_user_email: string;
    primary_user_full_name: string;
    computer_name: string;
    organization: string;
    is_unit_head: boolean;
}

export interface OrgRequest {
    id: number;
    contact_organization: string;
    application_name: string;
    family: string | null;
    family_version: string | null;
    family_edition: string | null;
    request: "add" | "remove";
    request_date: string;
    approved: boolean;
    completed: boolean;
    reviewed_by: string | null;
    reviewed_date: string | null;
    spc_id: number;
    requested_by: string;
}