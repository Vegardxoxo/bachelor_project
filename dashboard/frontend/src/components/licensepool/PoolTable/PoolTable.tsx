import * as React from 'react';
import {useEffect} from 'react';
import Box from '@mui/material/Box';
import Collapse from '@mui/material/Collapse';
import IconButton from '@mui/material/IconButton';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, {tableCellClasses} from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Typography from '@mui/material/Typography';
import Paper from '@mui/material/Paper';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import {LicensePoolData} from '../../../Interfaces';
import BuyButton from "../BuyButton/BuyButton";


interface RowProps {
    row: LicensePoolData;
}

interface Props {
    data: LicensePoolData[];
    handleSorting: (sortBy: string) => void;
}

/**
 * Custom row component for the LicensePool Table component
 * @param props - the row properties
 */
function Row(props: RowProps) {
    const {row} = props;
    const [open, setOpen] = React.useState(false);

    // Render a single row and its collapsible content
    return (
        <React.Fragment>
            { /* Main row content */}
            <TableRow sx={{'& > *': {borderBottom: 'unset'}}}>
                <TableCell>
                    <IconButton
                        aria-label="expand row"
                        size="small"
                        onClick={() => setOpen(!open)}
                    >
                        {open ? (
                            <KeyboardArrowUpIcon/>
                        ) : (
                            <KeyboardArrowDownIcon data-testid="KeyboardArrowDownIcon"/>
                        )}
                    </IconButton>

                </TableCell>
                <TableCell component="th" scope="row"
                           style={{textAlign: "left", fontFamily: 'Source Sans Pro,sans-serif'}}>
                    {row.application_name}
                </TableCell>
                <TableCell align="left" style={{
                    paddingRight: "20px",
                    fontFamily: 'Source Sans Pro,sans-serif'
                }}>{row.freed_by_organization}</TableCell>

                { /* Collapsible content row */}
            </TableRow>
            <TableRow>
                <TableCell style={{paddingBottom: 0, paddingTop: 0}} colSpan={6}>
                    <Collapse in={open} timeout="auto" unmountOnExit>
                        <Box sx={{margin: 1, marginLeft: '70px', paddingTop: '10px', paddingBottom: '10px'}}>
                            <Typography variant="h6" gutterBottom component="div"
                                        sx={{fontFamily: 'Source Sans Pro,sans-serif'}}>
                                {row.application_name}
                            </Typography>

                            { /* Sub-table with additional details */}
                            <Table size="small" aria-label="purchases" sx={{
                                [`& .${tableCellClasses.root}`]: {
                                    borderBottom: "none"
                                }
                            }}>
                                <colgroup>
                                    <col style={{width: '45%'}}/>
                                    <col style={{width: '35%'}}/>
                                    <col style={{width: '20%'}}/>
                                </colgroup>
                                <TableHead sx={{borderBottom: '#eee solid'}}>
                                    <TableRow>
                                        <TableCell align="left" sx={{
                                            fontFamily: 'Source Sans Pro,sans-serif',
                                            fontSize: '12pt',
                                            fontWeight: 800
                                        }}><b>Dato lagt til</b></TableCell>
                                        <TableCell align="left" sx={{
                                            fontFamily: 'Source Sans Pro,sans-serif',
                                            fontSize: '12pt',
                                            fontWeight: 800
                                        }}><b>Pris</b></TableCell>
                                        <TableCell align="left" sx={{
                                            fontFamily: 'Source Sans Pro,sans-serif',
                                            fontSize: '12pt',
                                            fontWeight: 800
                                        }}><b>Kj√∏p</b></TableCell>
                                    </TableRow>
                                </TableHead>
                                <TableBody>
                                    {row.details.map((detailRow) => (
                                        <TableRow key={detailRow.id}>
                                            <TableCell sx={{
                                                fontFamily: 'Source Sans Pro,sans-serif',
                                                fontSize: '12pt'
                                            }}>{detailRow.date_added ?? 'Ukjent'}</TableCell>
                                            <TableCell align="left" sx={{
                                                fontFamily: 'Source Sans Pro,sans-serif',
                                                fontSize: '12pt'
                                            }}>{detailRow.price ?? 500},-</TableCell>
                                            <TableCell> <BuyButton spc_id={detailRow.spc_id}
                                                                   application_name={row.application_name}
                                            />
                                            </TableCell>

                                        </TableRow>
                                    ))}
                                </TableBody>
                            </Table>
                        </Box>
                    </Collapse>
                </TableCell>
            </TableRow>
        </React.Fragment>
    );
}


/**
 * Main table component for the LicensePool page
 * @param data - the data to be displayed in the table
 * @param handleSorting - the function to be called when the user clicks a column header
 */
export default function PoolTable({data, handleSorting}: Props) {
    const software = data;
    const [loaded, setLoaded] = React.useState(false);

    // Set loaded state when data is received
    useEffect(() => {
        if ((software.length) > 0) {
            setLoaded(true);
        }

    }, [software]);

    // Render the main table
    return (
        <>
            {loaded ? (
                <div style={{width: "98.5%", marginLeft: '-0.4%'}}>
                    <TableContainer component={Paper}>
                        <Table aria-label="collapsible table">
                            <colgroup>
                                <col style={{width: '5%'}}/>
                                <col style={{width: '50%'}}/>
                                <col style={{width: '45%'}}/>
                            </colgroup>
                            <TableHead>
                                <TableRow>
                                    <TableCell style={{textAlign: "center"}}/>
                                    <TableCell onClick={() => handleSorting("application_name")}
                                               style={{
                                                   cursor: "pointer",
                                                   textAlign: "left",
                                                   fontFamily: 'Source Sans Pro,sans-serif',
                                                   fontSize: '12pt',
                                                   fontWeight: 800
                                               }}>
                                        <b>Lisensnavn &#9660;</b>
                                    </TableCell>
                                    <TableCell onClick={() => handleSorting("freed_by_organization")} align={"left"}
                                               style={{
                                                   cursor: "pointer",
                                                   textAlign: "left",
                                                   fontFamily: 'Source Sans Pro,sans-serif',
                                                   fontSize: '12pt'
                                               }}>
                                        <b>Frigitt av &#9660;</b>
                                    </TableCell>
                                </TableRow>
                            </TableHead>

                            <TableBody>
                                {software.map((user, index) => (
                                    <Row key={index} row={user}/>
                                ))}
                            </TableBody>
                        </Table>
                    </TableContainer>
                </div>
            ) : <h3>Velg programvare </h3>}
        </>
    );
}
