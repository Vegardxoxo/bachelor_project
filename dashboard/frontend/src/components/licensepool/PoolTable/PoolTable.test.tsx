import React from 'react';
import '@testing-library/jest-dom';
import {cleanup, render, screen} from "@testing-library/react";
import PoolTable from "./PoolTable";
import {LicensePoolData} from "../../../Interfaces";
import userEvent from "@testing-library/user-event";
import handleSorting from '../LicensePool';
import {RecoilRoot} from "recoil";
import renderer from "react-test-renderer";

const mockData: LicensePoolData[] = [
    {
        application_name: "Blackfish IE Tab 15",
        freed_by_organization: "IT-tjenesten",
        details: [
            {
                id: 1,
                freed_by_organization: "IT-tjenesten",
                application_name: "Blackfish IE Tab 15",
                date_added: "2023-04-08",
                family: "Blackfish IE Tab",
                family_version: "101500",
                family_edition: "100",
                price: 750,
                spc_id: 520185
            }
        ]
    }
];

describe('The pool table', () => {
    const handleSortingMock = jest.fn();

    beforeEach(() => {
        render(<RecoilRoot><PoolTable data={mockData} handleSorting={handleSortingMock}/></RecoilRoot>);
    })
    afterEach(() => {
        cleanup()
    })

    // Test if the component renders without crashing
    it('renders without crashing', async () => {
        expect(await screen.findByText('Lisensnavn ▼')).toBeInTheDocument();
        expect(await screen.findByText('IT-tjenesten')).toBeInTheDocument();
    })

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <RecoilRoot><PoolTable data={mockData} handleSorting={handleSortingMock}/></RecoilRoot>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })

    it('can expand and display expected details', async () => {
        const input = screen.getByTestId('KeyboardArrowDownIcon');
        expect(input).toBeInTheDocument();
        userEvent.click(input);
        expect(await screen.findByText('Dato lagt til')).toBeInTheDocument();
        expect(await screen.findByText('Pris')).toBeInTheDocument();
        expect(await screen.findByText('Kjøp')).toBeInTheDocument();
    })

    it('calls handleSorting when column header is clicked', async () => {
        const columnHeaders = await screen.findAllByText('Lisensnavn ▼');
        const columnHeader = columnHeaders[0]
        userEvent.click(columnHeader);
        expect(handleSortingMock).toHaveBeenCalledTimes(1);
        expect(handleSortingMock).toHaveBeenCalledWith('application_name');
    })

    // Test if the table renders correctly when data is empty
    it('renders empty message when data is empty', () => {
        render(<RecoilRoot><PoolTable data={[]} handleSorting={handleSorting}/></RecoilRoot>);
        expect(screen.getByText('Velg programvare')).toBeInTheDocument();
    })
})
