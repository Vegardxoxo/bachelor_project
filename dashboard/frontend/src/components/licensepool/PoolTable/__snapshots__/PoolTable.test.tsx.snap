// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The pool table matches snapshot 1`] = `
<h3>
  Velg programvare 
</h3>
`;
